### DISCUSSION TITLE
Add additional metrics to the vcpkg.json to help with C++20 modules adoption

### DISCUSSION THREAD

**Discussion by kelteseth** [microsoft/vcpkg#38513] [Ideas]
Hi 👋,

I'm the developer of [arewemodulesyet.org](https://arewemodulesyet.org/) that heavily relies on vcpkg to track the most popular open source packages (based on update count of the portfile). Now, there are some things I would like to see added to the vcpkg.json that would make at least my life easier:

- Used build system and the required version range
- Used c++ standard version range
- Level of modules adoption:
    1.  `LibraryX` cannot be used in a modules project (Often due to bugs in the compiler, for example [see this issue](https://github.com/stephenberry/glaze/issues/858))
    2. `LibraryX` can be used in a modules project
    3. `LibraryX` itself is a modules project
    4.  All of `LibraryX` dependencies are modules
    5. `LibraryX` uses C++23 import std;
    6.  All of `LibraryX` dependencies are using C++23 import std;
- Default C++ import statement for every library, we already track that here https://github.com/kelteseth/arewemodulesyet/blob/master/data/progress_overwrite.yml#L17

Cheers,
Eli

**Comment by dg0yt** [comment]
> to track the most popular open source packages (based on update count of the portfile).

Not convincing.

> Now, there are some things I would like to see added to the vcpkg.json that would make at least my life easier

Who is going to supply this information? There is already insufficient validation of updates, insufficient validation of installation order control and exported usage, insufficient validation of supported platforms and insufficient upstreaming of patches. Plus occasional circumventing of upstream build systems, making vpckg non-representative for the given ports. 

However, once you publish a well researched database, it might work the other way: port maintainers might be willing to check if the vcpkg port doesn't lag behind.

**Reply by kelteseth** [reply]
> Not convincing.

Do you have a better way we could track this? https://github.com/microsoft/vcpkg/discussions/17795 Has been open for about 6 years now...

> Who is going to supply this information? 

Well, we? The goal of the arewemodulesyet projects is to get the modules support rolling and the first thing I would do is to check if

A. My build system supports modules
B. My compiler supports modules
C. My dependencies support modules <- We are here

I do not want to move the work to you, but I think vcpkg would honestly be the best place to specify these informations.

**Comment by Neumann-A** [comment]
Just take the CI file list and check if the required files for module support are there?

**Reply by dg0yt** [reply]
These file lists are in the artifacts of the runs at https://dev.azure.com/vcpkg/public/_build?definitionId=29&_a=summary.

**Reply by kelteseth** [reply]
@Neumann-A not sure how I would check for modules support here. For example fmt:

VCPKGHeadersDatabase.txt
```
fmt:x64-windows:fmt/args.h
fmt:x64-windows:fmt/chrono.h
fmt:x64-windows:fmt/color.h
fmt:x64-windows:fmt/compile.h
fmt:x64-windows:fmt/core.h
fmt:x64-windows:fmt/format-inl.h
fmt:x64-windows:fmt/format.h
fmt:x64-windows:fmt/locale.h
fmt:x64-windows:fmt/os.h
fmt:x64-windows:fmt/ostream.h
fmt:x64-windows:fmt/printf.h
fmt:x64-windows:fmt/ranges.h
fmt:x64-windows:fmt/xchar.h
```

VCPKGDatabase.txt
```
fmt:x64-windows:/bin/fmt.dll
fmt:x64-windows:/bin/fmt.pdb
fmt:x64-windows:/debug/bin/fmtd.dll
fmt:x64-windows:/debug/bin/fmtd.pdb
fmt:x64-windows:/debug/lib/fmtd.lib
fmt:x64-windows:/debug/lib/pkgconfig/fmt.pc
fmt:x64-windows:/include/fmt/args.h
fmt:x64-windows:/include/fmt/chrono.h
fmt:x64-windows:/include/fmt/color.h
fmt:x64-windows:/include/fmt/compile.h
fmt:x64-windows:/include/fmt/core.h
fmt:x64-windows:/include/fmt/format-inl.h
fmt:x64-windows:/include/fmt/format.h
fmt:x64-windows:/include/fmt/locale.h
fmt:x64-windows:/include/fmt/os.h
fmt:x64-windows:/include/fmt/ostream.h
fmt:x64-windows:/include/fmt/printf.h
fmt:x64-windows:/include/fmt/ranges.h
fmt:x64-windows:/include/fmt/xchar.h
fmt:x64-windows:/lib/fmt.lib
fmt:x64-windows:/lib/pkgconfig/fmt.pc
fmt:x64-windows:/share/fmt/copyright
fmt:x64-windows:/share/fmt/fmt-config-version.cmake
fmt:x64-windows:/share/fmt/fmt-config.cmake
fmt:x64-windows:/share/fmt/fmt-targets-debug.cmake
fmt:x64-windows:/share/fmt/fmt-targets-release.cmake
fmt:x64-windows:/share/fmt/fmt-targets.cmake
fmt:x64-windows:/share/fmt/usage
fmt:x64-windows:/share/fmt/vcpkg_abi_info.txt
```

**Reply by Neumann-A** [reply]
Maybe read:
https://learn.microsoft.com/en-us/cpp/cpp/modules-cpp?view=msvc-170
and
https://learn.microsoft.com/en-us/cpp/build/compare-inclusion-methods?view=msvc-170

After that you can probably answer the question yourself

**Reply by kelteseth** [reply]
I do not see any module, so nope. Please enlightened me.

**Reply by Neumann-A** [reply]
> I do not see any module, so nope. Please enlightened me.

I see you didn't read it:
`The .ixx extension is required for module interface files in Visual Studio`

No files with `.ixx` installed -> no module support within vcpkg.
